### PR TITLE
(refactor) Refactor conditions form validation logic

### PR DESCRIPTION
**Pull request by denniskigen**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [ ] My work includes tests or is validated by existing tests.

## Summary

This PR refactors the conditions form's validation logic to use the zod schema as the source of truth instead of manual validation logic in the	`onSubmit` handler.

### CODE REVIEW

File: packages/esm-patient-conditions-app/src/conditions/conditions-form.workspace.tsx [hunk 1]
```diff
@@ -34,14 +25,26 @@ const ConditionsForm: React.FC<ConditionFormProps> = ({
...
+  const schema = z.object({
+    abatementDateTime: z.date().optional().nullable(),
+    clinicalStatus: z.string().refine((clinicalStatus) => formContext !== 'creating' || !!clinicalStatus, {
+      message: t('clinicalStatusRequired', 'A clinical status is required'),
+    }),
+    conditionName: z.string().refine((conditionName) => formContext !== 'creating' || !!conditionName, {
+      message: t('conditionRequired', 'A condition is required'),
+    }),
+    onsetDateTime: z.date().nullable(),
+  });
```
Comment (denniskigen) [inline packages/esm-patient-conditions-app/src/conditions/conditions-form.workspace.tsx:37]: This is the main change.

File: packages/esm-patient-conditions-app/src/conditions/conditions-form.workspace.tsx [hunk 2]
```diff
@@ -52,7 +55,7 @@ const ConditionsForm: React.FC<ConditionFormProps> = ({
             : null
           : null,
       conditionName: '',
-      clinicalStatus: condition?.cells?.find((cell) => cell?.info?.header === 'clinicalStatus')?.value ?? '',
+      clinicalStatus: formContext === 'editing' ? matchingCondition?.clinicalStatus?.toLowerCase() : '',
```
Comment (denniskigen) [inline packages/esm-patient-conditions-app/src/conditions/conditions-form.workspace.tsx:58]: The `clinicalStatus` value wasn't correctly getting picked when editing a condition. This fixes that.

File: packages/esm-patient-conditions-app/src/conditions/conditions-widget.component.tsx [hunk 3]
```diff
@@ -18,6 +18,8 @@ import {
 } from '@carbon/react';
 import { WarningFilled } from '@carbon/react/icons';
 import { useFormContext, Controller } from 'react-hook-form';
+// eslint-disable-next-line @typescript-eslint/consistent-type-imports
+import { z } from 'zod';
```
Comment (denniskigen) [inline packages/esm-patient-conditions-app/src/conditions/conditions-widget.component.tsx:22]: I can't get this to work sans the `eslint-disable` comment.

File: packages/esm-patient-conditions-app/src/conditions/conditions-form.workspace.tsx [hunk 4]
```diff
@@ -63,33 +66,14 @@ const ConditionsForm: React.FC<ConditionFormProps> = ({
...
-          type: 'manual',
-          message: t('clinicalStatusRequired', 'A clinical status is required'),
-        });
-      }
-      setIsSubmittingForm(false);
-    }
-
+  const onSubmit: SubmitHandler<z.infer<typeof schema>> = () => {
     setIsSubmittingForm(true);
   };
```
Comment (denniskigen) [inline packages/esm-patient-conditions-app/src/conditions/conditions-form.workspace.tsx:78]: This is the original validation logic.

File: packages/esm-patient-conditions-app/src/conditions/conditions-form.workspace.tsx [hunk 5]
```diff
@@ -102,6 +86,7 @@ const ConditionsForm: React.FC<ConditionFormProps> = ({
     <FormProvider {...methods}>
       <Form className={styles.form} onSubmit={methods.handleSubmit(onSubmit, onError)}>
         <ConditionsWidget
+          schema={schema}
```
Comment (denniskigen) [inline packages/esm-patient-conditions-app/src/conditions/conditions-form.workspace.tsx:89]: I'm now passing down this type annotation as props. It doesn't feel like the best approach.

File: packages/esm-patient-conditions-app/src/conditions/conditions-widget.component.tsx [hunk 6]
```diff
@@ -29,10 +31,15 @@ import {
...
-import { type ConditionSchema } from './conditions-form.workspace';
 import styles from './conditions-form.scss';
 
 interface ConditionsWidgetProps {
+  schema: z.ZodObject<{
+    abatementDateTime: z.ZodNullable<z.ZodOptional<z.ZodDate>>;
+    clinicalStatus: z.ZodEffects<z.ZodString, string, string>;
+    conditionName: z.ZodEffects<z.ZodString, string, string>;
+    onsetDateTime: z.ZodNullable<z.ZodDate>;
+  }>;
```
Comment (denniskigen) [inline packages/esm-patient-conditions-app/src/conditions/conditions-widget.component.tsx:42]: I feel that there's a better annotation than this or that can be inferred from usage.

File: packages/esm-patient-conditions-app/src/conditions/conditions-widget.component.tsx [hunk 7]
```diff
@@ -110,20 +118,17 @@ const ConditionsWidget: React.FC<ConditionsWidgetProps> = ({
     };
 
     try {
-      const res = await createCondition(payload);
+      await createCondition(payload);
+      mutate();
 
-      if (res.status === 201) {
```
Comment (denniskigen) [inline packages/esm-patient-conditions-app/src/conditions/conditions-widget.component.tsx:115]: This check isn't necessary per recent advice from @ibacher.

File: packages/esm-patient-conditions-app/src/conditions/conditions-widget.component.tsx [hunk 8]
```diff
@@ -256,7 +261,9 @@ const ConditionsWidget: React.FC<ConditionsWidgetProps> = ({
                   </ResponsiveWrapper>
                 )}
               />
-              {errors?.conditionName && <p className={styles.errorMessage}>{errors?.conditionName?.message}</p>}
+              {errors?.conditionName && (
+                <p className={styles.errorMessage}>{errors?.conditionName?.message as string}</p>
```
Comment (denniskigen) [inline packages/esm-patient-conditions-app/src/conditions/conditions-widget.component.tsx:265]: I imagine the `as string` type casting is necessitated by not using the most ideal type annotations.

**Comment by denniskigen** [comment]
@ibacher looks good?

**Review by ibacher — APPROVED** [review]

**Comment by denniskigen** [comment]
Thanks!

**Pull request merged**